tests/federation_room_join_partial_state_test.go

```diff
@@ -2405,6 +2405,77 @@ func TestPartialStateJoin(t *testing.T) {
 			mustSyncUntilDeviceListsHas(t, alice, syncToken, "left", server.UserID("elsie"))
 			mustQueryKeysWithFederationRequest(t, alice, userDevicesChannel, server.UserID("elsie"))
 		})
+
+		// test that device lists stop being tracked when leaving a partial state room before the
+		// partial state join completes.
+		t.Run("Device list no longer tracked when leaving partial state room", func(t *testing.T) {
+			t.Skip("Cannot yet leave a room during resync")
+
+			alice, server, userDevicesChannel, room, _, cleanup := setupDeviceListCachingTest(t, deployment, "t34alice")
+			defer cleanup()
+
+			// The room starts with @charlie and @derek in it.
+
+			// @t34alice:hs1 joins the room.
+			psjResult := beginPartialStateJoin(t, server, room, alice)
+			defer psjResult.Destroy()
+
+			syncToken := getSyncToken(t, alice)
+
+			// @elsie joins the room.
+			joinEvent := createJoinEvent(t, server, room, server.UserID("elsie"))
+			room.AddEvent(joinEvent)
+			server.MustSendTransaction(t, deployment, "hs1", []json.RawMessage{joinEvent.JSON()}, nil)
+			awaitEventViaSync(t, alice, room.RoomID, joinEvent.EventID(), syncToken)
+
+			// @elsie's device list ought to be cached.
+			syncToken = mustSyncUntilDeviceListsHas(t, alice, syncToken, "changed", server.UserID("elsie"))
+			mustQueryKeysWithFederationRequest(t, alice, userDevicesChannel, server.UserID("elsie"))
```

Comment:
this seems to be checking that the device list is *not* cached, in contrast to the comment?

Reply:
The first query (this one) brings the device list into the cache, and the second one tests that the device list is in the cache.

Reply:
so, when it says `// @elsie's device list ought to be cached.` ... we expect the device list to not be cached at that point? It doesn't seem very clear, either way.

Reply:
How does this sound?
```go
// @elsie's device list ought to be cached after the first request.
// Try to bring their device list into the homeserver's cache, then test whether it really was cached.
mustQueryKeysWithFederationRequest(...)
mustQueryKeysWithoutFederationRequest(...)
```

Reply:
Took me a moment to grok this as well. As I understand it, both `mustQueryKeys` functions query keys via the CSAPI. The `WithFederationRequest` checks that the server under test calls out over federation to the complement-managed server; the `Without` variant checks that this does not happen.

How about this?

```
The server under test should now be tracking @elsie's device list. Enforce this in two steps.
First, have Alice request Elsie's keys via the CS API and check that hs1 makes a federation request to serve Alice's request.
Then repeat Alice's request and check that hs1 does _not_ make a second federation request.
```

Reply:
I've had a go at rewriting this in af6228a. Please let me know what you think.

Reply:
yeah, this seems much clearer now.